var/spack/repos/builtin/packages/charm/package.py

```diff
@@ -92,6 +92,25 @@ class Charm(Package):
     variant("production", default=True, description="Build charm++ with all optimizations")
     variant("tracing", default=False, description="Enable tracing modules")
 
+    # FIXME: backend=mpi also provides mpi, but this can't be expressed in
+    # spack currently
```

Comment:
Can you make this more specific?  I think you're saying it can't be expressed b/c there would be two MPI providers in that configuration, and we don't allow two providers of the same virtual dependency in the same DAG.  Is that right?

Just for my own edification, how *do* you handle that case?  If the backend is MPI and Charm++ is providing MPI, how do you avoid symbol conflicts at runtime?

Reply:
I tried to clarify the comment a bit. The error message I get when writing `depends_on("mpi")` and `provides("mpi")` in the package is `==> Error: Cannot depend on 'charm@6.8.2%gcc@7.3.0 backend=mpi build-target=LIBS ~cuda~omp~papi+production~pthreads+shared+smp~syncft~tcp~tracing arch=linux-ubuntu18.04-x86_64' twice`.

In Charm++/AMPI, we rename symbols from `MPI_Foo` to `AMPI_Foo` in the user application when using MPI as the backend, which avoids the symbol conflict.